src/lib.rs

```diff
@@ -258,6 +266,185 @@ pub fn recv_msg(sock: &TsnSocket, msg: &mut msghdr) -> Result<isize, String> {
+            libc::SO_TIMESTAMPING,
+            &ts_flags as *const _ as *const libc::c_void,
+            std::mem::size_of::<u32>().try_into().unwrap(),
+        )
+    };
+    if err < 0 {
+        return Err(Error::last_os_error());
+    }
+
+    // setsoockopt for err queue
+
+    let flags: i32 = 1;
+    let err = unsafe {
+        libc::setsockopt(
+            sockfd,
+            libc::SOL_SOCKET,
+            libc::SO_SELECT_ERR_QUEUE,
+            &flags as *const _ as *const libc::c_void,
+            std::mem::size_of::<i32>().try_into().unwrap(),
+        )
+    };
+    if err < 0 {
+        return Err(Error::last_os_error());
+    }
+
+    // ioctl
+    let mut ts_cfg = libc::hwtstamp_config {
+        tx_type: libc::HWTSTAMP_TX_ON as i32,
+        rx_filter: libc::HWTSTAMP_FILTER_NONE as i32,
+        flags: 0,
+    };
+
+    let mut ifr_name: [libc::c_char; libc::IFNAMSIZ] = [0; libc::IFNAMSIZ];
+    for (source, target) in interface_name.as_bytes().iter().zip(ifr_name.iter_mut()) {
+        *target = *source as libc::c_char;
+    }
+
+    let ifreq = libc::ifreq {
+        ifr_name,
+        ifr_ifru: libc::__c_anonymous_ifr_ifru {
+            ifru_data: (&mut ts_cfg as *mut _) as *mut libc::c_char,
+        },
+    };
+
+    let err = unsafe {
+        // Not useless conversion because aarch64 has different type
+        #[allow(clippy::useless_conversion)]
+        libc::ioctl(sockfd, libc::SIOCSHWTSTAMP.try_into().unwrap(), &ifreq)
+    };
+    if err < 0 {
+        // XXX: While ioctl failed, SW timestamp is still enabled.
+        eprintln!("ioctl SIOCSHWTSTAMP failed: {}", Error::last_os_error());
+        eprintln!("But SW timestamp by kernel is still enabled.")
+    }
+
+    Ok(())
+}
+
+pub fn get_tx_timestamp(sock: &TsnSocket) -> Result<time::Timespec, Error> {
+    let sockfd = sock.fd;
+
+    let buf: [u8; 256] = [0u8; 256];
+    let buflen = std::mem::size_of_val(&buf);
+
+    let control: [u8; 256] = [0u8; 256];
+
+    let iov = libc::iovec {
+        iov_base: buf.as_ptr() as *mut libc::c_void,
+        iov_len: buflen,
+    };
+
+    let msg: libc::msghdr = unsafe {
+        // Avoid private field not provided error
+        let mut msg: libc::msghdr = std::mem::MaybeUninit::zeroed().assume_init();
+
+        msg.msg_name = std::ptr::null_mut();
+        msg.msg_namelen = 0;
+        msg.msg_iov = &iov as *const _ as *mut libc::iovec;
+        msg.msg_iovlen = 1;
+        msg.msg_control = control.as_ptr() as *mut libc::c_void;
+        msg.msg_controllen = {
+            // aarch64 has msg_controllen as u32, not usize
+            #[allow(clippy::useless_conversion)]
+            std::mem::size_of_val(&control).try_into().unwrap()
+        };
+        msg.msg_flags = 0;
+        msg
+    };
+
+    let pfd = libc::pollfd {
+        fd: sockfd,
+        events: libc::POLLPRI,
+        revents: 0,
+    };
+
+    let res = unsafe { libc::poll(&pfd as *const _ as *mut libc::pollfd, 1, 1000) };
```

Comment:
It would be better to change like this:

`let mut pfd = ...`

`let res = unsafe { libc::poll(&mut pfd, 1, 1000) }`

Reply:
I see. that's why `unsafe` block means unsafe literally